### PR TITLE
DetailsPage - pass extra resources to PageHeading Actions

### PR DESCRIPTION
**Pull request by suomiy**
It allows passing of additional resources to actions or to action creator. Needed for https://github.com/openshift/console/pull/1906

### CODE REVIEW

**Comment by suomiy** [comment]
@spadgett could you please take a look?

File: frontend/__tests__/components/utils/page-heading.spec.tsx [hunk 1]
```diff
@@ -60,7 +60,7 @@ describe(PageHeading.displayName, () => {
 
   it('renders breadcrumbs if given `breadcrumbsFor` function', () => {
     const breadcrumbs = [];
-    wrapper = wrapper.setProps({breadcrumbsFor: () => breadcrumbs, obj: {data: testResourceInstance}});
+    wrapper = wrapper.setProps({breadcrumbsFor: () => breadcrumbs, obj: { data: testResourceInstance, loaded:true, loadError: null }});
```
Comment (vojtechszocs) [inline frontend/__tests__/components/utils/page-heading.spec.tsx]: Nit: `loaded:true` missing space after `:`

File: frontend/public/components/factory/details.tsx [hunk 2]
```diff
@@ -2,37 +2,45 @@ import * as React from 'react';
...
+    } as FirehoseResource].concat(props.resources || [])}>
+      <PageHeading
+        detail={true}
+        title={props.name}
+        titleFunc={props.titleFunc}
+        menuActions={props.menuActions}
+        buttonActions={props.buttonActions}
+        kind={props.kind}
+        breadcrumbsFor={props.breadcrumbsFor ? props.breadcrumbsFor : breadcrumbsForDetailsPage(props.kindObj, props.match)}
+        resourceKeys={resourceKeys}
```
Comment (vojtechszocs) [inline frontend/public/components/factory/details.tsx]: This `resourceKeys` prop addition seems to be the only effective change here.

File: frontend/public/components/utils/kebab.tsx [hunk 3]
```diff
@@ -207,7 +207,9 @@ export type KebabOption = {
   href?: string, callback?: () => any;
   accessReview?: AccessReviewResourceAttributes;
 };
-export type KebabAction = (kind: K8sKind, obj: K8sResourceKind) => KebabOption;
+export type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption;
```
Comment (vojtechszocs) [inline frontend/public/components/utils/kebab.tsx]: Just curious, is `resources` expected to be an object? (Instead of an array?)

**Comment by vojtechszocs** [comment]
Just one nit-pick and question, otherwise LGTM.

@spadgett FYI, this PR touches core `DetailsPage` and `PageHeading` components.

File: frontend/public/components/utils/kebab.tsx [hunk 4]
```diff
@@ -207,7 +207,9 @@ export type KebabOption = {
...
+
+export type KebabOptionsCreator = (kindObj: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption[];
```
Comment (vojtechszocs) [inline frontend/public/components/utils/kebab.tsx]: @suomiy @rawagner @spadgett 

`KebabAction` is actually the "creator" of `KebabOption` (single) object, while the proposed `KebabOptionsCreator` is the "creator" of `KebabOption` object array, reusing the same argument list.

This adds ambiguity as to which one to use, since it's either a list of functions (`KebabAction[]`) or a single function (`KebabOptionsCreator`), each of which have different return types (and therefore different signatures), and therefore need to be handled differently.

We can make this simpler, without introducing an alternative type/signature

```ts
export type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption | KebabOption[];
```

and in `PageHeading` component

```tsx
const hasMenuActions = !_.isEmpty(menuActions); // no change here, contract stays the same
// { hasMenuActions && <ActionsMenu actions={menuActions.map(a => a(kindObj, data))} /> }
{ hasMenuActions && <ActionsMenu actions={menuActions.reduce((items: KebabOption[], action: KebabAction) => [
  ...items,
  ...(_.castArray(action(kindObj, data))),
], [])} /> }
```

This way, we use a single `KebabAction` type while supporting multiple outputs, i.e. `KebabOption | KebabOption[]`.

File: frontend/public/components/utils/kebab.tsx (hunk 3 above)
Comment (suomiy) [reply frontend/public/components/utils/kebab.tsx]: it could be literary anything

File: frontend/public/components/utils/kebab.tsx (hunk 4 above)
Comment (suomiy) [reply frontend/public/components/utils/kebab.tsx]: Yup, that would be nicer but it would not work.

The issue with this approach

- the resources are not the actual resources but are computed from the loaded data, please see for example vm status https://github.com/openshift/console/pull/1906/files#diff-e1ea60ed2e820073acdfd59753ed6636R147
- the resources are first accessible in PageHeading (please see DetailsPage)
- we have to compute something from these resources, hence the need for callback

**Comment by suomiy** [comment]
fixed the nit

File: frontend/public/components/utils/kebab.tsx (hunk 4 above)
Comment (vojtechszocs) [reply frontend/public/components/utils/kebab.tsx]: Both `KebabAction` and `KebabOptionsCreator` are functions with same arguments

```
kind: K8sKind, obj: K8sResourceKind, resources?: any
```

but different return value types, `KebabOption` vs. `KebabOption[]`.

The argument semantics when calling one function vs. the other one should be the same, right?

```js
_.isFunction(menuActions)
  ? menuActions(kindObj, data, extraResources) // case A
  : menuActions.map(a => a(kindObj, data, extraResources)) // case B
```

- In case A, we run one function and expect `KebabOption[]`.
- In case B, we run N functions each returning `KebabOption` and map result to `KebabOption[]`.
- The discriminator is the `menuActions` shape, either a single function (case A) or array of functions (case B).

---

**TL;DR**

In both `PageHeadingProps` and `DetailsPageProps`:

```diff
-menuActions?: any[];
+menuActions?: Function[] | KebabMultiAction; // FIXME: should be "KebabAction[] | KebabMultiAction"
```

At the bottom of `headings.tsx` (this type is unrelated to `kebab.tsx`, we need it specifically for `PageHeading`):

```ts
export type KebabMultiAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption[];
```

**Comment by vojtechszocs** [comment]
One more thing, TypeScript allows you to extract & reuse function parameter list, so you don't have to duplicate it:

```ts
type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption;
type KebabMultiAction = (...args: Parameters<KebabAction>) => KebabOption[];
```

**Comment by suomiy** [comment]
> One more thing, TypeScript allows you to extract & reuse function parameter list, so you don't have to duplicate it:
> 
> ```ts
> type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption;
> type KebabMultiAction = (...args: Parameters<KebabAction>) => KebabOption[];
> ```

good to know, changed

I am not convinced that `KebabMultiAction` is a good name though. It might imply that this is just one action which does multiple things.

**Comment by vojtechszocs** [comment]
> I am not convinced that `KebabMultiAction` is a good name though. It might imply that this is just one action which does multiple things.

`KebabAction` itself is a creator of `KebabOption` objects.

`KebabMultiAction` was supposed to imply it's a creator of multiple (array of) `KebabOption` objects.

Anyway, I'm not good at naming things, it was just a suggestion.

File: frontend/public/components/factory/details.tsx [hunk 5]
```diff
@@ -44,7 +52,7 @@ export type DetailsPageProps = {
   match: match<any>;
   title?: string | JSX.Element;
   titleFunc?: (obj: K8sResourceKind) => string | JSX.Element;
-  menuActions?: any[];
+  menuActions?: Function[] | KebabOptionsCreator; // FIXME shoulde be "KebabAction[] |" refactor pipeline-actions.tsx, etc.
```
Comment (vojtechszocs) [inline frontend/public/components/factory/details.tsx]: Typo: `should be`

File: frontend/public/components/utils/headings.tsx [hunk 6]
```diff
@@ -104,8 +105,9 @@ export type PageHeadingProps = {
   detail?: boolean;
   kind?: K8sResourceKindReference;
   kindObj?: K8sKind;
-  menuActions?: any[];
-  obj?: {data: K8sResourceKind};
+  menuActions?: Function[] | KebabOptionsCreator; // FIXME shoulde be "KebabAction[] |" refactor pipeline-actions.tsx, etc.
```
Comment (vojtechszocs) [inline frontend/public/components/utils/headings.tsx]: Typo: `should be`

File: frontend/public/components/utils/headings.tsx [hunk 7]
```diff
@@ -134,3 +136,4 @@ PageHeading.displayName = 'PageHeading';
 ResourceOverviewHeading.displayName = 'ResourceOverviewHeading';
 SectionHeading.displayName = 'SectionHeading';
 SidebarSectionHeading.displayName = 'SidebarSectionHeading';
+
```
Comment (vojtechszocs) [inline frontend/public/components/utils/headings.tsx]: Why the extra newline?

File: frontend/public/components/utils/kebab.tsx [hunk 8]
```diff
@@ -207,7 +207,9 @@ export type KebabOption = {
   href?: string, callback?: () => any;
   accessReview?: AccessReviewResourceAttributes;
 };
-export type KebabAction = (kind: K8sKind, obj: K8sResourceKind) => KebabOption;
+export type KebabAction = (kind: K8sKind, obj: K8sResourceKind, resources?: any) => KebabOption;
+
+export type KebabOptionsCreator = (...args: Parameters<KebabAction>) => KebabOption[];
```
Comment (vojtechszocs) [inline frontend/public/components/utils/kebab.tsx]: As suggested in my other comment, I'd prefer to put this type next to React component that actually uses it, i.e. in `public/components/utils/headings.tsx`.

File: frontend/public/components/utils/headings.tsx (hunk 7 above)
Comment (suomiy) [reply frontend/public/components/utils/headings.tsx]: oops

File: frontend/public/components/utils/kebab.tsx (hunk 8 above)
Comment (suomiy) [reply frontend/public/components/utils/kebab.tsx]: ok, moved to `headings.tsx`
